lib/spack/spack/compilers/__init__.py

```diff
@@ -125,6 +125,105 @@ def get_compiler_config(scope=None, init_config=True):
     return config
 
 
+def get_compiler_config_from_packages(scope=None):
+    """Return the compiler configuration from packages.yaml"""
+    packages = []
+    config = spack.config.get("packages", scope=scope)
+
+    if not config:
+        return packages
+
+    for name, entry in config.items():
+        if name not in supported_compilers() and name not in package_name_to_compiler_name.keys():
+            continue
+        packages.extend(_compiler_config_from_package_config(entry))
+
+    return packages
+
+
+def _compiler_config_from_package_config(config):
+    externals = config.get("externals", None)
+    if not externals:
+        return []
+
+    compilers = []
+    for entry in externals:
+        compiler = _compiler_config_from_external(entry)
+        if compiler:
+            compilers.append(compiler)
+
+    return compilers
+
+
+def _compiler_config_from_external(config):
+    spec = spack.spec.Spec(config["spec"])
+    compiler_spec = spack.spec.CompilerSpec(
+        package_name_to_compiler_name.get(spec.name, spec.name), spec.version
+    )
+
+    extra_attributes = config.get("extra_attributes", {})
+    prefix = config.get("prefix", None)
+
+    compiler_class = class_for_compiler_name(compiler_spec.name)
+    paths = extra_attributes.get("paths", {})
+    compiler_langs = ["cc", "cxx", "fc", "f77"]
+    for compiler in compiler_langs:
+        if compiler in paths:
+            continue
+
+        if not prefix:
+            continue
+
+        bindir = prefix
+        if os.path.basename(prefix) != "bin":
+            bindir = os.path.join(prefix, "bin")
+
+        # Check for files that satisfy the naming scheme for this compiler
+        for file, regexp in itertools.product(
+            os.listdir(bindir), compiler_class.search_regexps(compiler)
+        ):
+            match = regexp.match(file)
+            if match:
+                paths[compiler] = os.path.join(bindir, file)
```

Comment:
Can we require to find the attributes in `packages`? These are very few lines, so there is that, but it is effectively the third way to search for compilers (`compiler find`, `external find`, this to fill manually added spec in externals)

Reply:
I don't think we can require `paths` to be present in `packages.yaml` to consider an entry -- it flies in the face of the purpose of this PR to move towards seamlessly using compilers like other dependencies.

We can write this same detection logic individually into each compiler package, and then call it from there, but that doesn't seem like much of a win. We're already using the class-specific search regex list for each compiler.

I think if we need to move this somewhere else, the most reasonable option is to move it to a standalone method in this module, and eventually move it to a `CompilerPackage` mixin that the compiler packages inherit from.

Does that fit with what you were suggesting/satisfy your interests?

Reply:
> I don't think we can require paths to be present in packages.yaml to consider an entry -- it flies in the face of the purpose of this PR to move towards seamlessly using compilers like other dependencies.

Getting back to this after starting reviewing #43464 Above we didn't notice that the attributes we set in externals are not named `paths` and `fc`, `f77`. That was because we have extra code that is "smart" enough to recompute some paths. It would have been better, imo, to fail and detect the error earlier. Said otherwise, this logic masked a bug :slightly_frowning_face: